domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -261,37 +340,342 @@ class TopicListController @Inject constructor(
+    // The list of topic IDs that are considered "finished" from a recommendation perspective.
+    val fullyCompletedTopicIds = topicProgressList.filter {
+      topicHasAtLeastOneStoryCompleted(it)
+    }.map(TopicProgress::getTopicId)
+    // A set of topic IDs that can be considered topics that should not be recommended.
+    val impliedFinishedTopicIds = computeImpliedCompletedDependencies(
+      fullyCompletedTopicIds, topicDependencyMap
+    )
+    // Suggest prerequisite topic user needs to learn after completing any of the topics.
+    // The order in which the topic IDs are enumerated matters, and that it should be in the order
+    // of the list itself.
+    for (topicId in unstartedTopicIdList) {
+      // All of the topic's prerequisites can be suggested if the topic is ongoing.
+      val dependentTopicIds = topicDependencyMap[topicId] ?: listOf()
+      if (topicId !in impliedFinishedTopicIds &&
+        impliedFinishedTopicIds.containsAll(dependentTopicIds)
+      ) {
+        createRecommendedStoryFromAssets(topicId)?.let {
+          recommendedStories.add(it)
+        }
+      }
+    }
+    return recommendedStories
+  }
+
+  private fun topicHasAtLeastOneStoryCompleted(it: TopicProgress): Boolean {
+    val topic = topicController.retrieveTopic(it.topicId)
+    return it.storyProgressMap.values.any { storyProgress ->
+      val story = topicController.retrieveStory(topic.topicId, storyProgress.storyId)
+      return@any checkIfStoryIsCompleted(storyProgress, story)
+    }
+  }
+
+  /**
+   * Return the list of topic IDs that are completed or can be implied completed based on actually
+   * completed topics.
+   */
+  private fun computeImpliedCompletedDependencies(
+    fullyCompletedTopicIds: List<String>,
+    topicDependencyMap: Map<String, Set<String>>
+  ): Set<String> {
+    // For each completed topic ID, compute the transitive closure of all of its dependencies &
+    // then combine them into a single list with the actual completed topic IDs. The returned list
+    // is a list of either completed or assumed completed topics which will eliminate potential
+    // recommendations.
+    val completedTopicIds =
+      fullyCompletedTopicIds.flatMap { topicId ->
+        computeTransitiveDependencyClosure(topicId, topicDependencyMap)
+      } + fullyCompletedTopicIds
+    return completedTopicIds.toSet()
+  }
+
+  private fun computeTransitiveDependencyClosure(
+    topicId: String,
+    topicDependencyMap: Map<String, Set<String>>
+  ): Set<String> {
+    // Compute the total list of dependent topics that must be completed before the specified topic
+    // can be recommended. Note that this will cause a stack overflow if the graph has cycles.
+    val topicDependencyList = topicDependencyMap[topicId]?.flatMap { dependentId ->
+      computeTransitiveDependencyClosure(
+        dependentId,
+        topicDependencyMap
+      )
+    }
+    return topicDependencyList?.toSet() ?: emptySet()
```

Comment:
```suggestion
    val directDependencies = topicDependencyMap[topicId] ?: listOf()
    val transitiveDependencies = directDependencies.flatMap { dependentId ->
      computeTransitiveDependencyClosure(
        dependentId,
        topicDependencyMap
      )
    }
    return (transitiveDependencies + directDependencies).toSet()
```

I think the bug causing the closure to be computed incorrectly is we never actually include the direct dependencies (which means the recursion yields no results since the base case isn't starting with the start nodes in the topic graph).